src/vm/gccover.cpp

```diff
@@ -1293,7 +1293,7 @@ bool IsGcCoverageInterrupt(LPVOID ip)
 // original instruction. Only one instruction must be used, 
 // because multiple threads can be executing the same code stream.
 
-void RemoveGcCoverageInterrupt(Volatile<BYTE>* instrPtr, BYTE * savedInstrPtr)
+void RemoveGcCoverageInterrupt(TADDR instrPtr, BYTE * savedInstrPtr)
```

Comment:
Do the two writes to instrPtr actually need to be Volatile writes so that other threads don't keep using old values?
If so the old code is still wrong,but we coudl use:

` *(Volatile<WORD>*)instrPtr  = *(WORD *)savedInstrPtr;`

Reply:
Perhaps `RemoveGcCoverageInterrupt `is running single threaded and there is a memory barrier in `FlushInstructionCache `before we we resume multi-threaded code.

Reply:
The comment in the header indicates that the update must be atomic, which AFAIK require a properly aligned `instrPtr` value

Reply:
@briansull On arm32, all instructions are aligned on 2 byte boundaries (the bit 0 is the thumb bit indicating which mode the processor runs in). As for arm64, all instruction addresses are 4 byte aligned since all instructions are 4 bytes long. So the alignment is naturally ensured.

Reply:
As for memory barriers, it seems they are not needed since even if we put a memory barrier here, there would be no guarantee that other processors would see the change in a bound time. Memory barrier would just ensure that operations before and after the barrier are seen by other processors as happening before or after the barrier. Only the instruction cache flush can ensure that other processors will see the instruction change right away. 
But in this case, I think the we really don't care. The thing is that if there are two or more processors running the same piece of code, they are competing for storing the invalid instructions and restoring the original ones. So we might restore the original instruction here and another processor might patch it back to the invalid one right after. The result is that when this happens, some threads might perform the GC more than once at the same instruction and some threads might not do the GC at some instructions.

Reply:
On arm32 in thumb mode you have a mix of 2 byte and 4 byte instructions.
You can have a 4 byte instruction that is not aligned on a 4 byte boundary.
Such an instruction can even cross two cache lines.

Reply:
@briansull I am sorry, I've forgotten that we also patch the 4 byte instructions by 4 byte versions of the illegal instructions. But we cannot really do anything about the 4 byte alignment, since we need to be able to patch instructions at any arbitrary address aligned to 2 bytes.